resilience4j-retrofit/src/test/java/io/github/resilience4j/retrofit/RetrofitRetryTest.java

```diff
@@ -0,0 +1,232 @@
+            .connectTimeout(TIMEOUT, TimeUnit.MILLISECONDS)
+            .readTimeout(TIMEOUT, TimeUnit.MILLISECONDS)
+            .writeTimeout(TIMEOUT, TimeUnit.MILLISECONDS)
+            .build();
+
+        this.service = new Retrofit.Builder()
+            .addCallAdapterFactory(RetryCallAdapter.of(retry))
+            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
+            .baseUrl(wireMockRule.baseUrl())
+            .client(client)
+            .build()
+            .create(RetrofitService.class);
+    }
+
+    @Test
+    public void shouldNotRetrySuccessfulCall() throws Exception {
+        stubFor(get(urlPathEqualTo("/greetingsResponse"))
+            .willReturn(aResponse()
+                .withStatus(200)
+                .withHeader("Content-Type", "text/plain")
+                .withBody("hello world")));
+
+        service.greetingsResponse().execute();
+
+        verify(1, getRequestedFor(urlPathEqualTo("/greetingsResponse")));
+    }
+
+    @Test
+    public void shouldNotRetrySuccessfulEnqueuedCall() throws Throwable {
+        stubFor(get(urlPathEqualTo("/greetingsResponse"))
+            .willReturn(aResponse()
+                .withStatus(200)
+                .withHeader("Content-Type", "text/plain")
+                .withBody("hello world")));
+
+        EnqueueDecorator.enqueue(service.greetingsResponse());
+
+        verify(1, getRequestedFor(urlPathEqualTo("/greetingsResponse")));
+    }
+
+
+    @Test(expected = IOException.class)
+    public void shouldRetryOnIOExceptionTimingOutCall() throws IOException {
+        stubFor(get(urlPathEqualTo("/greetingsResponse"))
+            .willReturn(aResponse()
+                .withFixedDelay(500)
+                .withStatus(200)
+                .withHeader("Content-Type", "text/plain")
+                .withBody("hello world")));
+
+        try {
+            service.greetingsResponse().execute();
+        } catch (Exception e) {
+            throw new IOException();
+        }
+
+        final Retry.Metrics metrics = retry.getMetrics();
+        verify(3, getRequestedFor(urlPathEqualTo("/greetingsResponse")));
+        assertThat(metrics.getNumberOfSuccessfulCallsWithRetryAttempt())
+            .isEqualTo(0);
+    }
+
+    @Test(expected = IOException.class)
+    public void shouldRetryOnIOExceptionForCancelledCall() throws IOException {
+        stubFor(get(urlPathEqualTo("/greetingsResponse"))
+            .willReturn(aResponse()
+                .withFixedDelay(500)
+                .withStatus(200)
+                .withHeader("Content-Type", "text/plain")
+                .withBody("hello world")));
+
+        try {
+            Call<ResponseBody> call = service.greetingsResponse();
+            cancelAsync(call, 100);
+            call.execute();
+        } catch (Exception e) {
+            throw new IOException();
+        }
+
+        verify(3, getRequestedFor(urlPathEqualTo("/greetingsResponse")));
```

Comment:
Is it really expected that the call is retried 3 times, even if the call is cancelled?

Reply:
Yes, @RobWin. As I have added IO Exception in retry config and the call is throwing IO Exception (Socket Timeout due to call being canceled). Thus, it'll retry.
`
private static final RetryConfig retryConfig = RetryConfig<Response<String>>custom().maxAttempts(3).retryOnResult(response -> response.code() == 504).intervalFunction(IntervalFunction.ofExponentialBackoff()).retryExceptions(IOException.class).build();`